docs/proposals/hpa-custom-metrics.md

```diff
@@ -0,0 +1,162 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<!-- TAG RELEASE_LINK, added by the munger automatically -->
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.3/docs/proposals/custom-metrics.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+Specifying Metrics in the HPA
+=============================
+
+Introduction
+------------
+
+Currently, custom metrics specified via a list of metric names
+and target values.  Each metric name is prefixed with "custom/",
+and then is requested at the pod level for each pod of the target
+scalable.
+
+Several use cases are not supported by this design:
+
+### Refering to Non-Custom Metrics ###
+
+Several users have asked for the ability to scale on existing built-in metrics
+(e.g. memory usage).  While the current design does allow the user to specify
+a metric name, all metrics are prefixed with "custom/", meaning built-in
+metrics cannot be passed this way.  A new design should not have completely
+separate mechanisms for custom and non-custom metrics.  This also allows
+refering to raw CPU usage rate, instead of refering to CPU usage as a
+percentage of the CPU request (this has been a commonly asked-for feature
```

Comment:
> but I'm strongly against deleting option to define CPU utilization as a percentage of request.

Did you see the comment below about adding it back in?  It's missing from this proposal mainly by oversight, but I'm not against keeping it in.

> Why is it easier to express CPU in raw usage?

There's two factors:

1. By far, the most common "why isn't my HPA autoscaling working" issue that I've seen (between GitHub issues for Kubernetes and OpenShift, OpenShift IRC, Kubernetes Slack, and other sources) has ended up being that the user forgot to set request on one of their containers.  I've also had multiple people ask me why they couldn't just set the value directly, and what "percentage of request" meant, and how it related to percentage of a CPU core.

2. Some cluster setup automatically set requests and limits based on user account information, etc.  When the request is automatically set, asking the user to set a percentage of request is a bit strange to the user.  Furthermore, if the request is set to something fairly low (so that pods are not guaranteed many resources, and you get many pods per node, the target percentage ends up needing to be something in the hundreds of percents or more, which is odd.